rollup/sync_service/sync_service.go

```diff
@@ -0,0 +1,179 @@
+package sync_service
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+	"time"
+
+	"github.com/scroll-tech/go-ethereum/core/rawdb"
+	"github.com/scroll-tech/go-ethereum/ethdb"
+	"github.com/scroll-tech/go-ethereum/log"
+	"github.com/scroll-tech/go-ethereum/node"
+	"github.com/scroll-tech/go-ethereum/params"
+)
+
+// DefaultFetchBlockRange is the number of blocks that we collect in a single eth_getLogs query.
+const DefaultFetchBlockRange = uint64(20)
+
+// DefaultPollInterval is the frequency at which we query for new L1 messages.
+const DefaultPollInterval = time.Second * 15
+
+// DbWriteThresholdBytes is the size of batched database writes in bytes.
+const DbWriteThresholdBytes = 10 * 1024
+
+// DbWriteThresholdBlocks is the number of blocks scanned after which we write to the database
+// even if we have not collected DbWriteThresholdBytes bytes of data yet. This way, if there is
+// a long section of L1 blocks with no messages and we stop or crash, we will not need to re-scan
+// this secion.
+const DbWriteThresholdBlocks = 100
+
+// SyncService collects all L1 messages and stores them in a local database.
+type SyncService struct {
+	ctx                  context.Context
+	cancel               context.CancelFunc
+	client               *BridgeClient
+	db                   ethdb.Database
+	pollInterval         time.Duration
+	latestProcessedBlock uint64
+}
+
+func NewSyncService(ctx context.Context, genesisConfig *params.ChainConfig, nodeConfig *node.Config, db ethdb.Database, l1Client EthClient) (*SyncService, error) {
+	// terminate if the caller does not provide an L1 client (e.g. in tests)
+	if l1Client == nil || (reflect.ValueOf(l1Client).Kind() == reflect.Ptr && reflect.ValueOf(l1Client).IsNil()) {
+		log.Warn("No L1 client provided, L1 sync service will not run")
+		return nil, nil
```

Comment:
need return a error.New("xxx") ?

Reply:
We allow `nil` (i.e. no sync service) since many tests instantiate a node.